### DISCUSSION TITLE
[Chapter 07] Advice 애너테이션에 pointcut 표현식 설정

### DISCUSSION THREAD

**Discussion by coalong** [cowaine/spring-boot-for-msa#76] [Chapter 07. Spring AOP, Testing, and Auto Confiugration] [answered]
P.393 에서 어드바이스 애너테이션에 포인트컷 설정하는 방식이 설명되어있다.

아래 같은 문장이 있는데, 

> 참고로 @AfterReturning, @AfterThrowing 애너 테이션은 value와 같은 포인트 컷 속성을 추가로 제공한다. 
  

5가지 `@Around`, `@Before`, `@AfterReturning`, `@AfterThrowing`, `@After` 중에서 
왜 `@AfterReturning`, `@AfterThrowing` 만 value 외에 pointcut 속성이 있을까? (value 와 같은 역할)

<img width="724" alt="image" src="https://github.com/cowaine/spring-boot-for-msa/assets/71637466/a6e5afbf-ef01-4da7-a125-258048bb49eb">

**Comment by BvrPark** [comment, marked as answer]
위에서 첨부하신 그림에서 보면
**value**는 어떤 조인 포인트에 어드바이스를 연결하는데 사용되는 포인트 컷 표현식을 지정한다.
**pointcut**어떤 조인 포인트에 어드바이스를 연결하는데 사용되는 포인트 컷 표현식을 지정하는데,
**pointcut**이 지정될 경우에는 **value**설정을 대체한다.라고 나와있습니다.

그래서 **value**는 간단한 포인트 컷 표현식 지정 
**pointcut**는 보다 세세한 포인트 컷 표현식 지정에 사용된다고 생각이 드는데

`@Before`이나 `@After` 같은 경우에는 어드바이스의 실행시점이 이미 정해져 있기 때문에 세세하게 지정할 필요 없이 간단하게 포인트 컷 지정을 해도 되서 **pointcut**이라는 속성이 없는 것 같고

`@Around`는 **pointcut**대신에 메서드의 파라미터로 `ProceedingJoinPoint`를 가져와서 조인포인트를 제어하기 때문에 **pointcut**이라는 속성이 없는 것 같습니다. 

`@AfterReturning`은 메서드가 성공적으로 실행되고 난 후에 어드바이스를 실행하고
`@AfterThrowing`은 메서드가 예외를 던진 후에 어드바이스를 실행하고 추가로 **throwing**속성을 사용해서 던진 예외도 다시 바인딩합니다.
이렇게 두 어노테이션은 메서드의 실행 결과에 따라서 동작을 수행해야 되므로 보다 세밀한 조인 포인트 지정이 필요한 **pointcut**속성이 존재하는 것 같습니다.

</br>

### 참고자료
[https://stackoverflow.com/questions/48955420/afterreturningpointcut-vs-afterreturningvalue](https://stackoverflow.com/questions/48955420/afterreturningpointcut-vs-afterreturningvalue)

**Comment by Irisation23** [comment]
### 금주 CheckPoint

- `@AfterReturning`, `@AfterThrowing` 애너 테이션은 value와 같은 포인트 컷 속성을 추가로 제공한다.
- value와 pointcut 에 대한 차이 추가적으로 정리가 필요해 보임
